### PR TITLE
Draft: Update pyth and withdraw

### PR DESCRIPTION
**Pull request by Babybluess**

### CODE REVIEW

File: lending_contract/sources/loan.move [hunk 1]
```diff
@@ -33,6 +39,12 @@ module lending_contract::loan {
     const BORROWER_PAID_STATUS: vector<u8> = b"BorrowerPaid";
     const FINISHED_STATUS: vector<u8> = b"Finished";
 
+    const DEFAULT_RATE_FACTOR: u64 = 10000;
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:42]: duplicated

File: lending_contract/sources/loan.move [hunk 2]
```diff
@@ -33,6 +39,12 @@ module lending_contract::loan {
...
+    const SECOND_IN_YEAR: u64 = 31536000;
+
+    const SUI_COIN_TYPE: vector<u8> = b"0x587c29de216efd4219573e08a1f6964d4fa7cb714518c2c8a0f29abfa264327d";
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:45]: redundant?

File: lending_contract/sources/loan.move [hunk 3]
```diff
@@ -131,7 +173,8 @@ module lending_contract::loan {
 
         let collateral_amount = coin::value<T2>(&collateral);
     
-        assert!(is_valid_collateral(configuration, lend_amount, collateral_amount), ECollateralNotValidToMinHealthRatio);
+        assert!(is_valid_collateral(lend_amount, configuration, collateral_amount, 0,
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:176]: format code, too long

File: lending_contract/sources/loan.move [hunk 4]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
         balance::split<T2>(&mut loan.collateral, amount)
     }
 
-    fun is_valid_collateral(
-        configuration: &Configuration,
+    public entry fun is_valid_collateral(
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:412]: is_valid_collateral_amount
or
check_health_ratio

File: lending_contract/sources/loan.move [hunk 5]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
         lend_amount: u64,
+        configuration: &Configuration,
         collateral_amount: u64,
+        withdraw_amount: u64,
+        clock: &Clock,
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:417]: comment: clock should move to end, before context (ctx)

File: lending_contract/sources/loan.move [hunk 6]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
+        pyth_state: &PythState,
+        price_info_object_collateral: &PriceInfoObject,
+        price_info_object_lending: &PriceInfoObject,
+        time_threshhold: u64,
+        ctx: &mut TxContext,
     ): bool {
-        //TODO: use price feeds getting price lend token price and collateral token price to check health ratio
-        true
+        let remaining_collateral_amount = collateral_amount - withdraw_amount;
+        let remaining_collateral_price_total = get_price_total(clock, pyth_state, price_info_object_collateral, remaining_collateral_amount, time_threshhold);
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:425]: get_value_by_usd

File: lending_contract/sources/loan.move [hunk 7]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
+
+        if( current_health_ratio < min_health_ratio) {
+            false
+        } else {
+            true
+        }
+
+    }
+
+    fun get_price_total (
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:438]: move all price function to other module, such as PriceFeed

File: lending_contract/sources/loan.move [hunk 8]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
+        price: &I64,  
+    ): u64 {
+        let price_value: u64;
+        let is_negative = i64::get_is_negative(price);
+        if (!is_negative) {
+            price_value = i64::get_magnitude_if_positive(price);
+        } else {
+            price_value = i64::get_magnitude_if_negative(price);
+        };
+        price_value
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:461]: what is unit of price_value here?
https://docs.pyth.network/price-feeds/best-practices

File: lending_contract/sources/loan.move [hunk 9]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
+        lender: address,
+        borrower: address,
+        clock: &Clock,
+        ctx: &mut TxContext,
+    ) {
+        let offer_key = offer::new_offer_key<T1>(offer_id);
+        assert!(state::contain<OfferKey<T1>, Offer<T1>>(state, offer_key), EOfferNotFound);
+        let offer = state::borrow_mut<OfferKey<T1>, Offer<T1>>(state, offer_key);
+
+        let _tier_id = offer::get_tier_id<T1>(offer);
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:521]: why not tier_id?

File: lending_contract/sources/loan.move (hunk 4 above)
Comment (pando-dreamer) [reply lending_contract/sources/loan.move:412]: hàm này để private thôi

File: lending_contract/sources/loan.move [hunk 10]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
         balance::split<T2>(&mut loan.collateral, amount)
     }
 
-    fun is_valid_collateral(
-        configuration: &Configuration,
+    public entry fun is_valid_collateral(
         lend_amount: u64,
+        configuration: &Configuration,
         collateral_amount: u64,
+        withdraw_amount: u64,
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:416]: mục tiêu là check health ratio, bỏ input withdraw amount

File: lending_contract/sources/loan.move [hunk 11]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
+        price_info_object_lending: &PriceInfoObject,
+        time_threshhold: u64,
+        ctx: &mut TxContext,
     ): bool {
-        //TODO: use price feeds getting price lend token price and collateral token price to check health ratio
-        true
+        let remaining_collateral_amount = collateral_amount - withdraw_amount;
+        let remaining_collateral_price_total = get_price_total(clock, pyth_state, price_info_object_collateral, remaining_collateral_amount, time_threshhold);
+        let lend_price_total = get_price_total(clock, pyth_state, price_info_object_lending, lend_amount, time_threshhold);
+        let current_health_ratio = remaining_collateral_price_total / lend_price_total;
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:427]: Cần tính cả decimals của từng loại coin ở đây

File: lending_contract/sources/loan.move [hunk 12]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
+            price_value = i64::get_magnitude_if_positive(price);
+        } else {
+            price_value = i64::get_magnitude_if_negative(price);
+        };
+        price_value
+    }
+
+     public entry fun withdraw_collateral_loan_offer<T1, T2>(
+        loan_id: ID,
+        offer_id: ID,
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:466]: No need offer_id here, can get from Loan data

File: lending_contract/sources/loan.move [hunk 13]
```diff
@@ -359,12 +409,157 @@ module lending_contract::loan {
...
+        configuration: &Configuration,
+        withdraw_amount: u64,
+        pyth_state: &PythState,
+        price_info_object_collateral: &PriceInfoObject,
+        price_info_object_lending: &PriceInfoObject,
+        time_threshhold: u64,
+        ctx: &mut TxContext,
+    ) {
+        let offer_key = offer::new_offer_key<T1>(offer_id);
+        assert!(state::contain<OfferKey<T1>, Offer<T1>>(state, offer_key), EOfferNotFound);
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:478]: should check loan is valid here than check offer

File: lending_contract/sources/loan.move [hunk 14]
```diff
@@ -114,6 +152,10 @@ module lending_contract::loan {
         collateral: Coin<T2>,
         lend_coin_metadata: &CoinMetadata<T1>,
         collateral_coin_metadata: &CoinMetadata<T2>,
+        pyth_state: &PythState,
+        price_info_object_collateral: &PriceInfoObject,
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:156]: Need to check PriceInfoObject is valid (for type T1/T2) or not

**Review by pando-dreamer — CHANGES_REQUESTED** [review]

File: lending_contract/sources/loan.move [hunk 15]
```diff
@@ -359,12 +432,245 @@ module lending_contract::loan {
...
+        let coin_symbol = string::from_ascii(coin_symbol_ascii);
+
+        if (coin_name == validated_name && coin_decimals == validated_decimals && coin_symbol == validated_symbol) {
+            true
+        } else {
+            false
+        }
+    }
+
+     public entry fun withdraw_collateral_loan_offer<T1, T2>(
```
[the remark this answers is not in the document]
Comment (pando-dreamer) [reply lending_contract/sources/loan.move:557]: - version, configuration, state
- loan_id, withdraw_amount --> loan
- coin_metadata
- pyth objects

File: lending_contract/sources/loan.move [hunk 16]
```diff
@@ -359,12 +432,245 @@ module lending_contract::loan {
...
+            loan_id,
+            borrower: loan.borrower,
+            withdraw_amount,
+            remaining_collateral_amount,
+            timestamp: current_timestamp,
+        });
+    }
+
+    public entry fun deposit_collateral_loan_offer<T1, T2>(
+        configuration: &Configuration,
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:613]: tương tự withdraw, thiếu version, thống nhất lại về thứ tự params

File: lending_contract/sources/loan.move [hunk 17]
```diff
@@ -359,12 +432,245 @@ module lending_contract::loan {
...
+        lend_coin_metadata: &CoinMetadata<T1>,
+        collateral_coin_metadata: &CoinMetadata<T2>,
+        loan_id: ID,
+        deposit_coin: Coin<T2>,
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:617]: deposit_amount

File: lending_contract/sources/loan.move [hunk 18]
```diff
@@ -359,12 +432,245 @@ module lending_contract::loan {
...
+        let ( asset_tier ) = {
+            let offer_key = offer::new_offer_key<T1>(offer_id);
+            assert!(state::contain<OfferKey<T1>, Offer<T1>>(state, offer_key), EOfferNotFound);
+            let offer = state::borrow<OfferKey<T1>, Offer<T1>>(state, offer_key);
+            offer::get_asset_tier(offer)
+        };
+
+        let loan = state::borrow_mut<LoanKey<T1, T2>, Loan<T1, T2>>(state, loan_key);
+
+        let _lend_fee_percentage = configuration::lender_fee_percent(configuration);
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:638]: lend_fee_percentage

File: lending_contract/sources/loan.move [hunk 19]
```diff
@@ -359,12 +432,245 @@ module lending_contract::loan {
...
+        let _borrow_fee_percentage = configuration::borrower_fee_percent(configuration);
+
+        let lend_token_ascii = coin::get_symbol<T1>(lend_coin_metadata);
+        let lend_token = string::from_ascii(lend_token_ascii);
+        let collateral_token_ascii = coin::get_symbol<T2>(collateral_coin_metadata);
+        let collateral_token = string::from_ascii(collateral_token_ascii);
+
+        let current_timestamp = clock::timestamp_ms(clock);
+        let _status = string::utf8(MATCHED_STATUS);
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:647]: check contract status: only in status FUND_TRANSFERRED

File: lending_contract/sources/loan.move [hunk 20]
```diff
@@ -359,12 +432,245 @@ module lending_contract::loan {
...
+        ctx: &mut TxContext,
+    ) {
+        let usdc_name = string::utf8(USDC_NAME);
+        let usdc_symbol = string::utf8(USDC_SYMBOL);
+        let sui_name = string::utf8(SUI_NAME);
+        let sui_symbol = string::utf8(SUI_SYMBOL);
+
+        let loan_key = new_loan_key<T1, T2>(loan_id);
+        let current_timestamp = clock::timestamp_ms(clock);
+        let loan = state::borrow_mut<LoanKey<T1, T2>, Loan<T1, T2>>(state, loan_key);
```
Comment (pando-dreamer) [inline lending_contract/sources/loan.move:577]: need check contract status

File: lending_contract/sources/loan.move (hunk 19 above)
Comment (pando-dreamer) [reply lending_contract/sources/loan.move:647]: get status contract

File: lending_contract/sources/utils.move [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+    friend lending_contract::offer;
+    friend lending_contract::loan;
+    friend lending_contract::price_feed;
+
+    public(friend) fun power(base: u64, exponent: u64): u64 {
+        let result = 1;
+
+        if (exponent == 0) {
+            result
```
Comment (pando-dreamer) [inline lending_contract/sources/utils.move:12]: thừa

File: lending_contract/sources/loan.move (hunk 15 above)
[the remark this answers is not in the document]
Comment (pando-dreamer) [reply lending_contract/sources/loan.move:500]: Vẫn thiếu version, chưa check loan status

**Review by pando-dreamer — CHANGES_REQUESTED** [review]

File: lending_contract/sources/offer.move [hunk 22]
```diff
@@ -263,6 +269,12 @@ module lending_contract::offer {
         offer.lender
     }
 
+    public fun get_asset_tier<T>(
```
Comment (pando-dreamer) [inline lending_contract/sources/offer.move:272]: duplicate with get_tier_id?

File: lending_contract/sources/utils.move [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+            price_value = i64::get_magnitude_if_negative(price);
+        };
+        price_value
+    }
+
+    public (friend) fun is_negative (
+        price: &I64,
+    ): bool {
+        let is_negative = i64::get_is_negative(price);
+        if (!is_negative) {
```
Comment (pando-dreamer) [inline lending_contract/sources/utils.move:37]: return is_negative

File: lending_contract/sources/price_feed.move [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+        let exponent_power = utils::power(10, exponent_u64);
+        let price_usd: u128;
+
+        if (utils::is_negative(&price::get_expo(&price))) {
+            price_usd = ((price_u64 * max_decimals_power / exponent_power) as u128);
+        } else {
+            price_usd = ((price_u64 * max_decimals_power * exponent_power) as u128);
+        };
+        
+        let value_usd = price_usd * ((amount * utils::power(10, sub_decimals) / exponent_power) as u128);
```
Comment (pando-dreamer) [inline lending_contract/sources/price_feed.move:43]: wrong calculation

**Review by pando-dreamer — CHANGES_REQUESTED** [review]

File: lending_contract/sources/price_feed.move [hunk 25]
```diff
@@ -25,24 +26,21 @@ module lending_contract::price_feed {
...
         let price_u64 = utils::i64_to_u64(&price::get_price(&price));
         let exponent_u64 = utils::i64_to_u64(&price::get_expo(&price));
-        let max_decimals_power = utils::power(10, max_decimals);
         let exponent_power = utils::power(10, exponent_u64);
-        let price_usd: u128;
+        let value_usd: u128;
 
         if (utils::is_negative(&price::get_expo(&price))) {
-            price_usd = ((price_u64 * max_decimals_power / exponent_power) as u128);
+            value_usd = ((price_u64 * amount_by_max_decimals / exponent_power) as u128);
```
Comment (ducbtsn) [inline lending_contract/sources/price_feed.move:39]: may cause overflow here

File: lending_contract/sources/price_feed.move [hunk 26]
```diff
@@ -25,24 +26,21 @@ module lending_contract::price_feed {
     ) : u128 {
         let time_threshold = configuration::price_time_threshold(configuration);
         let coin_decimals_u8 = coin::get_decimals<T>(coinMetadata);
-        let sub_decimals = max_decimals - (coin_decimals_u8 as u64);
+        // Standardized amount to max decimals
+        let amount_by_max_decimals = amount * utils::power(10, max_decimals) / utils::power(10, (coin_decimals_u8 as u64));
```
Comment (ducbtsn) [inline lending_contract/sources/price_feed.move:30]: may cause overflow here

**Pull request merged**